src/layout/layout.rs

```diff
@@ -1254,30 +1640,30 @@ mod tests {
             test(Rect::new(0, 0, 1, 1), &[Max(0), Max(1)], "b"); // zero, exact
             test(Rect::new(0, 0, 1, 1), &[Max(0), Max(2)], "b"); // zero, overflow
             test(Rect::new(0, 0, 1, 1), &[Max(1), Max(0)], "a"); // exact, zero
-            test(Rect::new(0, 0, 1, 1), &[Max(1), Max(1)], "a"); // exact, exact
-            test(Rect::new(0, 0, 1, 1), &[Max(1), Max(2)], "a"); // exact, overflow
+            test(Rect::new(0, 0, 1, 1), &[Max(1), Max(1)], "b"); // exact, exact with stretchlast
+            test(Rect::new(0, 0, 1, 1), &[Max(1), Max(2)], "b"); // exact, overflow with stretchlast
```

Comment:
Lots more bad results. I know these are edge cases, but they tend to be intuitively hitting the wrong side of what I'd expect, which suggests that they will also make off by one decisions all over the place. "Centered" layout being 1 column to the right etc. I'm ommitting further comments about bad changes below this point. Also note, this is probably subjective, but I can't find an obvious justification for why these results should be preferred over the alternatives.

Reply:
In all of these bad results, what you are saying is "intuitive" doesn't match the what a user would expect from the name `StretchLast`. `StretchLast` is supposed to mean excess into last.

fwiw, if I add a new `test_with_stretch` like below, I get the previous behavior:

```rust
        #[track_caller]
        fn test_with_stretch(area: Rect, constraints: &[Constraint], expected: &str) {
            let layout = Layout::default()
                .direction(Direction::Horizontal)
                .constraints(constraints)
                .flex(Flex::Stretch) // solve with `Flex::Stretch`
                .split(area);
            let mut buffer = Buffer::empty(area);
            for (i, c) in ('a'..='z').take(constraints.len()).enumerate() {
                let s: String = c.to_string().repeat(area.width as usize);
                Paragraph::new(s).render(layout[i], &mut buffer);
            }
            let expected = Buffer::with_lines(vec![expected]);
            assert_buffer_eq!(buffer, expected);
        }
```

What you are calling intuitive is unfortunately an implementation defined behavior, and relying on that can lead to unstable solutions unfortunately.

My recommendation is either

- accept the new behavior as `StretchLast` (it is stable)
- OR make `Flex::Stretch` the default

Reply:
The issues are:
- this changes existing behavior, which means apps with painfully tweaked layouts now change and that annoys devs
- the behavior of stretch last doesn't imply last area is more important, it implies that this applies only once all other constraints have been met - I can't work out a way where you would describe how `Length(3), Length(2) -> "bb"` makes sense ever. The first area is bigger than the second, this should result in either "ab" or "aa", and never "bb".

> What you are calling intuitive is unfortunately an implementation defined behavior, and relying on that can lead to unstable solutions unfortunately.

This was stable previously though, so this is a regression without a way to fix right? Don't neither of your options result in behavior that matches the existing layout, or am I missing something?

Reply:
It wasn't a stable solution in general, it just happened to be reproducible for this particular test. There's no guarantee that for any `[Length(x), Length(y)]` or `[Length(x), Length(y), Length(z)]` or any number of constraints for any size of the rect that you'll get the same behavior. 

I also think using two or multiple `Length`s like this is unlikely? Users that have tweaked their layouts with this are depending on an implementation specific behavior. Similarly, we have tweaked the priorities for `Percentage` and `Ratio` in this PR but our tests do not capture all the different ways this could behave. I think commonly used patterns will be unaffected by this change in this PR.

> The first area is bigger than the second, this should result in either "ab" or "aa", and never "bb".

Can I ask you to think what should the following do in these 3 cases? (think about what a user coming to Ratatui completely new to layouts / terminal user interfaces etc would expect):

```rust
for width in [10, 50, 100] {
    let area = Rect(0, 0, width, 0);
    let (a1, a2) = area.split(Layout::horizontal([Length(30), Length(70)]))
}
```

fwiw, if you comment these lines in `StretchLast`, you'll get the old behavior:

```rust
if let Some((segment, _)) = segments
    .iter()
    .zip(&layout.constraints)
    .sorted_by(|(_, c1), (_, c2)| Ord::cmp(&c1.rank(), &c2.rank()))
    .next_back()
{
    solver.add_constraint(segment.grows_to(size))?;
}
```

If you do comment it, users will get unstable layouts. An example where that removing this block of code will result in this test pass / fail randomly:

```rust
#[case::previously_unstable_test(vec![50, 25, 25], vec![Length(25), Length(25), Fixed(25)])] 
#[case::previously_unstable_test(vec![25, 50, 25], vec![Length(25), Length(25), Fixed(25)])]
```

In my opinion, having a layout that is random and unpredictable like this is worse than a breaking change in behavior.

---

If we absolutely don't want to break existing user behavior, I'd recommend changing `Flex::StretchLast` to `Flex::Unstable` and removing the block of code from above, and working toward changing the default in the future.

Reply:
> My takeaway on that hypothetical is that intuitively we should consider lengths as being proportional when constraints cannot be met.

The problem is cassowary (or linear programming in general) cannot handle "conditionals". There's no way in the formulation to say "if this, do that". We have to preprocess those cases and selectively add different constraints if we want it to behave differently.

> is there a test that fails in 0.25 like this that would be a better example?

Here's an example of an unstable test on the v0.25.0 tag:

```rust
#[test]
fn unstable_test() {
    let area = Rect::new(0, 0, 100, 0);
    let chunks = Layout::default()
        .direction(Direction::Horizontal)
        .constraints([
            Constraint::Length(25),
            Constraint::Length(25),
            Constraint::Max(25),
        ])
        .split(area);
    dbg!(chunks.iter().map(|r| r.width).collect::<Vec<u16>>());
}
```

<img width="758" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/c852df11-36eb-4e7b-9650-9b7e56696d5a">

I think my point is that you are drawing "meaning" from these simple tests but that's not what you should do, because that's not how cassowary will behave all the time with the constraints we have in v0.25.0. In `v0.25.0` and in the current `main`, how the default solve behaves is undefined. And we've named that undefined behavior `LastTakesRemainder` / `StretchLast` because that's behavior we observed in simple test cases. But our tests are under specified. 

I think we should strive for solutions that are reproducible and stable, particularly more so because we have now included `Fixed` and `Proportional`.

Reply:
> I'd strongly want that to give stable [38,37,25] regardless of which flex option is chosen

This is kind of why I want `Stretch` to be the default, since it is more intuitive that what happens with `StretchLast` in this PR OR in `main`.

Reply:
While we can't have conditionals, doesn't this just highlight that there's probably a missing rule here though?
Let's constrain our discussion to an over-allocation that cannot be met: 70/30 or 30/70 into width 50:
Here's what that looks like from a cassowary perspective (I'm going to simplify this to ignore spacers, and just use
3 variables (in fact we only need 1 as everything else is fixed, but let's go with 3 for now - a, b, c (in the 0.25 real impl this would be a b c d with b = c = required)
1.  a = 0 (start) required
2. a <= b required
3. b <= c required
4. c = 50 (end) required
5. b - a = 70 strong
6. c - b = 30 strong
7. a >= 0, b >= 0, c >=0
8. a <=50, b <=50, c<=50

This doesn't have a solution that doesn't violate either 5 or 6, but we can add a nice default that should apply regardless of what flex type we choose:
(b - a) / 30 = (c-b / 70) medium

This is a gross simplification, but reasonable. This constraint only kicks in once the solver tries to break one of the more important constraints right? (This is basically the example in the main doc btw.)

---

Incidentally, this discussion suggests that it would be really useful to split the *split methods at the point where all the constraints are calculated, so that we can just test that part (and not from an add a `dbg!` macro approach, but an actual testable function approach - I think this might help visualize some of the odd edge case behavior better.

For that to work, we'd have to collect the constraints outside of the solver instead of directly in it as it has no getters. That's a simple but really useful thing that makes some of our talk earlier where should the call to solver.add_constraint belong lie kinda moot - it belongs we've created all the constraints.

Reply:
> This constraint only kicks in once the solver tries to break one of the more important constraints right? (This is basically the example in the main doc btw.)

I like that change, i.e. make `Length` behave more like `Proportional`. (All the more reason to get rid of `Proportional` then?)

Reply:
I pushed a new commit where `Length`, `Percentage` and `Ratio` are proportional.

https://github.com/ratatui-org/ratatui/pull/788/commits/5b7792fa7f6a72c9f13a7d666a75f0e482373f60

These tests still change, so it would be breaking the UI.

Reply:
I've reverted the change for these bad results. I'm resolving these conversations.